topology/nhlt/intel/dmic/dmic-process.c

```diff
@@ -0,0 +1,1379 @@
+	}
+
+	if (cfg->mfir_b > 0) {
+		cfg->fir_b_length = cfg->fir_b->length;
+		ret = fir_coef_scale(&cfg->fir_b_scale, &cfg->fir_b_shift,
+				     cfg->fir_b->shift, cfg->fir_b->coef,
+				     cfg->fir_b->length, gain_to_fir);
+		if (ret < 0) {
+			/* Invalid coefficient set found, should not happen. */
+			fprintf(stderr, "select_mode(): invalid coefficient set found");
+			return -1;
+		}
+	} else {
+		cfg->fir_b_scale = 0;
+		cfg->fir_b_shift = 0;
+		cfg->fir_b_length = 0;
+	}
+
+	return 0;
+}
+
+/* The FIFO input packer mode (IPM) settings are somewhat different in HW versions. This helper
+ * function returns a suitable IPM bit field value to use.
+ */
+static void ipm_helper1(struct intel_dmic_params *dmic, int *ipm)
+{
+	int di = dmic->dmic_dai_index;
+	int pdm[DMIC_HW_CONTROLLERS];
+	int i;
+
+	/* Loop number of PDM controllers in the configuration. If mic A or B is enabled then a pdm
+	 * controller is marked as active for this DAI.
+	 */
+	for (i = 0; i < DMIC_HW_CONTROLLERS; i++) {
+		if (dmic->dmic_prm[di].pdm[i].enable_mic_a ||
+		    dmic->dmic_prm[di].pdm[i].enable_mic_b)
+			pdm[i] = 1;
+		else
+			pdm[i] = 0;
+	}
+
+	/* Set IPM to match active pdm controllers. */
+	*ipm = 0;
+
+	if (pdm[0] == 0 && pdm[1] > 0)
+		*ipm = 1;
+
+	if (pdm[0] > 0 && pdm[1] > 0)
+		*ipm = 2;
+}
+
+static void ipm_helper2(struct intel_dmic_params *dmic, int source[], int *ipm)
+{
+	int di = dmic->dmic_dai_index;
+	int pdm[DMIC_HW_CONTROLLERS];
+	int i;
+	int n = 0;
+
+	for (i = 0; i < OUTCONTROLX_IPM_NUMSOURCES; i++)
+		source[i] = 0;
+
+	/* Loop number of PDM controllers in the configuration. If mic A or B is enabled then a pdm
+	 * controller is marked as active. The function returns in array source[] the indice of
+	 * enabled pdm controllers to be used for IPM configuration.
+	 */
+	for (i = 0; i < DMIC_HW_CONTROLLERS; i++) {
+		if (dmic->dmic_prm[di].pdm[i].enable_mic_a ||
+		    dmic->dmic_prm[di].pdm[i].enable_mic_b) {
+			pdm[i] = 1;
+			source[n] = i;
+			n++;
+		} else {
+			pdm[i] = 0;
+		}
+	}
+
+	/* IPM bit field is set to count of active pdm controllers. */
+	*ipm = pdm[0];
+	for (i = 1; i < DMIC_HW_CONTROLLERS; i++)
+		*ipm += pdm[i];
+}
+
+/* Loop number of PDM controllers in the configuration. The function checks if the controller should
+ * operate as stereo or mono left (A) or mono right (B) mode. Mono right mode is setup as channel
+ * swapped mono left.
+ */
+static int stereo_helper(struct intel_dmic_params *dmic, int stereo[], int swap[])
+{
+	int cnt;
+	int i;
+	int swap_check;
+	int ret = 0;
+
+	for (i = 0; i < DMIC_HW_CONTROLLERS; i++) {
+		cnt = 0;
+		if (dmic->dmic_prm[0].pdm[i].enable_mic_a ||
+		    dmic->dmic_prm[1].pdm[i].enable_mic_a)
+			cnt++;
+
+		if (dmic->dmic_prm[0].pdm[i].enable_mic_b ||
+		     dmic->dmic_prm[1].pdm[i].enable_mic_b)
+			cnt++;
+
+		/* Set stereo mode if both mic A anc B are enabled. */
+		cnt >>= 1;
+		stereo[i] = cnt;
+
+		/* Swap channels if only mic B is used for mono processing. */
```

Comment:
it would make it a lot easier to understand, if you had a check like:
```
if (stereo[i]) continue;
```
then the check for !cnt below becomes redundant

Reply:
talking to Seppo, original had many revisions about this, with several contributors -> so conclusion maybe not touch this this time.